### DISCUSSION TITLE
No tree-shaking ?

### DISCUSSION THREAD

**Discussion by esoubiran-aneo** [vmware-clarity/ng-clarity#402] [Ideas]
Hello,

Is there any way to have three-shaking when using Clarity?

![image](https://user-images.githubusercontent.com/103997512/202199248-82c4c3fc-9318-497e-90d6-0157ecd561dd.png)

**Comment by esoubiran-aneo** [comment]
Fresh Angular App
![image](https://user-images.githubusercontent.com/103997512/202207726-da3c7a9b-8d4c-4eca-8b4e-836d4e72992f.png)

Fresh Angular App Using Clarity 
![image](https://user-images.githubusercontent.com/103997512/202208091-47e04f12-c1a8-4080-9d90-9d4d8f59d9e9.png)

That's very heavy. Install done with this guide: https://clarity.design/documentation/get-started

Is there a way to reduce this?

```json
  "dependencies": {
    "@angular/animations": "^14.2.0",
    "@angular/common": "^14.2.0",
    "@angular/compiler": "^14.2.0",
    "@angular/core": "^14.2.0",
    "@angular/forms": "^14.2.0",
    "@angular/platform-browser": "^14.2.0",
    "@angular/platform-browser-dynamic": "^14.2.0",
    "@angular/router": "^14.2.0",
    "@cds/core": "^6.2.2",
    "@clr/angular": "^13.10.1",
    "@clr/ui": "^13.10.1",
    "rxjs": "~7.5.0",
    "tslib": "^2.3.0",
    "zone.js": "~0.11.4"
  },
````

**Comment by esoubiran-aneo** [comment]
Ok, its seems there is tree-shaking but components are so huge. Datagrid is 150ko

(First build is without datagrid, second is with empty datagrid)
![image](https://user-images.githubusercontent.com/103997512/202211513-5a6a8434-8c8b-4374-adcd-98e30b607235.png)

**Comment by esoubiran-aneo** [comment]
So, maximal budget is used just with a data grid and clarity, it seems very complicated to then build an app

**Comment by bbogdanov** [comment]
Just as a workaround until the team investigates that - You can use only the modules you need instead of importing the `ClarityModule`. e.g. `ClrDropdownModule` which only loads the `clr-dropdown` component and necessary services and components around it. 

Also it will be good if you share your `angular.json` production configuration for better understanding what the prod build does.

**Comment by esoubiran-aneo** [comment]
Here my angular.json.

```json
{
  "projectType": "application",
  "root": "apps/app",
  "sourceRoot": "apps/app/src",
  "prefix": "armonik.admin.gui",
  "targets": {
    "build": {
      "executor": "@angular-devkit/build-angular:browser",
      "outputs": [
        "{options.outputPath}"
      ],
      "options": {
        "outputPath": "dist/apps/app",
        "index": "apps/app/src/index.html",
        "main": "apps/app/src/main.ts",
        "polyfills": "apps/app/src/polyfills.ts",
        "tsConfig": "apps/app/tsconfig.app.json",
        "inlineStyleLanguage": "scss",
        "assets": [
          "apps/app/src/favicon.ico",
          "apps/app/src/assets"
        ],
        "styles": [
          "node_modules/@clr/icons/clr-icons.min.css",
          "node_modules/@clr/ui/clr-ui-dark.min.css",
          "apps/app/src/styles.scss"
        ],
        "scripts": []
      },
      "configurations": {
        "production": {
          "budgets": [
            {
              "type": "initial",
              "maximumWarning": "500kb",
              "maximumError": "1.2mb"
            },
            {
              "type": "anyComponentStyle",
              "maximumWarning": "2kb",
              "maximumError": "4kb"
            }
          ],
          "fileReplacements": [
            {
              "replace": "apps/app/src/environments/environment.ts",
              "with": "apps/app/src/environments/environment.prod.ts"
            }
          ],
          "outputHashing": "all"
        },
        "development": {
          "buildOptimizer": false,
          "optimization": false,
          "vendorChunk": true,
          "extractLicenses": false,
          "sourceMap": true,
          "namedChunks": true
        }
      },
      "defaultConfiguration": "production"
    },
    "serve": {
      "executor": "@angular-devkit/build-angular:dev-server",
      "configurations": {
        "production": {
          "browserTarget": "app:build:production"
        },
        "development": {
          "browserTarget": "app:build:development"
        }
      },
      "defaultConfiguration": "development",
      "options": {
        "proxyConfig": "apps/app/proxy.conf.json"
      }
    },
    "extract-i18n": {
      "executor": "@angular-devkit/build-angular:extract-i18n",
      "options": {
        "browserTarget": "app:build"
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": [
          "apps/app/src/**/*.ts",
          "apps/app/src/**/*.html"
        ]
      }
    },
    "test": {
      "executor": "@angular-devkit/build-angular:karma",
      "options": {
        "main": "apps/app/src/test.ts",
        "tsConfig": "apps/app/tsconfig.spec.json",
        "karmaConfig": "apps/app/karma.conf.js",
        "polyfills": "apps/app/src/polyfills.ts",
        "styles": [],
        "scripts": [],
        "assets": []
      }
    }
  },
  "tags": []
}

```

**Comment by esoubiran-aneo** [comment]
Is this the official doc to use icons in an angular app?

https://core.clarity.design/foundation/icons/

**Comment by kevinbuhmann** [comment]
> Is this the official doc to use icons in an angular app?
> 
> https://core.clarity.design/foundation/icons/

Yes.

**Comment by esoubiran-aneo** [comment]
> Just as a workaround until the team investigates that - You can use only the modules you need instead of importing the `ClarityModule`. e.g. `ClrDropdownModule` which only loads the `clr-dropdown` component and necessary services and components around it.
> 
> Also it will be good if you share your `angular.json` production configuration for better understanding what the prod build does.

Even with this solution ([can be found here](https://github.com/aneoconsulting/ArmoniK.Admin.GUI/pull/305)), it seems that every clarity components are loaded in the main. Will investigate more.

**Comment by Jinnie** [comment]
Real tree-shaking has never been part of the Clarity Angular design. It was a goal for the Clarity Core (web-component based) project, which is now on hold.

As I don't want to close this request, but there is nothing we can do about it now, I'll move it to the Discussions, where we can track it as feature request and keep discussing it there.

CC: @pdaigle

**Reply by boeckMt** [reply]
> ... It was a goal for the Clarity Core (web-component based) project, which is now on hold. ...

@Jinnie can you say why **Clarity Core is on hold now** and when will it continue?

We started rewriting our applications to use **Clarity Core**  if possible and try to prevent using **Clarity Angular**... or is that not a good decision?

**Comment by Jinnie** [comment]
We're experiencing some severe team reorganizations. We don't expect to be working on Core until our new team is completely structured. Currently, we don't have the resource to work on Core. We hope to be able to share some more specific information as soon as possible, but right now our focus is back to Clarity Angular and especially working on the Accessibility.
